python/pyspark/worker.py

```diff
@@ -74,17 +75,37 @@ def wrap_udf(f, return_type):
 
 
 def wrap_pandas_udf(f, return_type):
-    arrow_return_type = toArrowType(return_type)
-
-    def verify_result_length(*a):
-        result = f(*a)
-        if not hasattr(result, "__len__"):
-            raise TypeError("Return type of pandas_udf should be a Pandas.Series")
-        if len(result) != len(a[0]):
-            raise RuntimeError("Result vector from pandas_udf was not the required length: "
-                               "expected %d, got %d" % (len(a[0]), len(result)))
-        return result
-    return lambda *a: (verify_result_length(*a), arrow_return_type)
+    if isinstance(return_type, StructType):
+        import pyarrow as pa
+
+        arrow_return_types = list(to_arrow_type(field.dataType) for field in return_type)
```

Comment:
I prefer list comprehensions, but it's up to you

Reply:
Fixed.